packages/client/lib/sync/fetcher/blockfetcher.ts

```diff
@@ -85,13 +59,7 @@ export class BlockFetcher extends Fetcher {
     blocks = blocks.map((b: Block) =>
       Block.fromValuesArray(b.raw(), { common: this.config.common })
     )
-    await this.chain.blockchain.initPromise
-    for (let i = 0; i < blocks.length; i++) {
-      const block: Block = blocks[i]
-
-      await this.chain.blockchain.putBlock(block)
-      await this.vm.runBlockchain()
-    }
+    await this.chain.blockchain.putBlocks(blocks)
```

Comment:
I don't think this is correct to tear this apart. Otherwise the chain will be updated without knowing if the blocks actually pass in the VM.

Reply:
Any opinions here @cgewecke @ryanio ? 🤔

Reply:
Note: if we put blocks in the chain (say 128 blocks of the Eth classic chain while we are on mainnet), then after this `update` will be called and this will invoke the `runBlocks` of `fullsync`. This will call `runBlockchain`. If the block is a bad block, then `runBlockchain` actually **deletes** this block from the blockchain. When a block is deleted from the chain, also **all children** will be removed. So, if it is a bad block, then all of these 128 blocks will be removed, and the `while` loop in fullsync will also halt at some point since the head hash will not be updated.

The problem is though (maybe this was your point) that we want to report that the peer reporting these blocks is reporting bad blocks and we should thus disconnect and ban this peer.